File: solutions/go/complex-numbers/1/complex_numbers.go
package complexnumbers

import "math"

type Number struct{ re, im float64 }

func (n Number) Real() float64 { return n.re }

func (n Number) Imaginary() float64 { return n.im }

func (a Number) Add(b Number) Number {
	return Number{re: a.re + b.re, im: a.im + b.im}
}

func (a Number) Subtract(b Number) Number {
	return Number{re: a.re - b.re, im: a.im - b.im}
}

func (a Number) Multiply(b Number) Number {
	return Number{re: a.re*b.re - a.im*b.im, im: a.re*b.im + a.im*b.re}
}

func (n Number) Times(f float64) Number {
	return Number{re: n.re * f, im: n.im * f}
}

func (a Number) Divide(b Number) Number {
	d := b.re*b.re + b.im*b.im
	return Number{
		re: (a.re*b.re + a.im*b.im) / d,
		im: (a.im*b.re - a.re*b.im) / d,
	}

}

func (n Number) Conjugate() Number { return Number{re: n.re, im: -n.im} }

func (n Number) Abs() float64 { return math.Sqrt(n.re*n.re + n.im*n.im) }

func (n Number) Exp() Number {
	f := math.Exp(n.re)
	return Number{re: f * math.Cos(n.im), im: f * math.Sin(n.im)}
}

File: solutions/haskell/complex-numbers/1/src/ComplexNumbers.hs
module ComplexNumbers (
    Complex,
    conjugate,
    abs,
    exp,
    real,
    imaginary,
    mul,
    add,
    sub,
    div,
    complex,
) where

import Prelude hiding (abs, div, exp)
import Prelude qualified

-- Data definition -------------------------------------------------------------

data Complex a = (:+) {real, imaginary :: a} deriving (Eq, Show)
infix 6 :+

complex :: (a, a) -> Complex a
complex = uncurry (:+)

-- unary operators -------------------------------------------------------------

conjugate :: (Num a) => Complex a -> Complex a
conjugate (a :+ b) = a :+ (-b)

abs :: (Floating a) => Complex a -> a
abs (a :+ b) = sqrt $ sq a + sq b

exp :: (Floating a) => Complex a -> Complex a
exp (a :+ b) = eᵃ * cos b :+ eᵃ * sin b where eᵃ = Prelude.exp a

-- binary operators ------------------------------------------------------------

mul :: (Num a) => Complex a -> Complex a -> Complex a
mul (xᵣ :+ xᵢ) (yᵣ :+ yᵢ) = (xᵣ * yᵣ - xᵢ * yᵢ) :+ (xᵣ * yᵢ + xᵢ * yᵣ)

add :: (Num a) => Complex a -> Complex a -> Complex a
add (xᵣ :+ xᵢ) (yᵣ :+ yᵢ) = (xᵣ + yᵣ) :+ (xᵢ + yᵢ)

sub :: (Num a) => Complex a -> Complex a -> Complex a
sub (xᵣ :+ xᵢ) (yᵣ :+ yᵢ) = (xᵣ - yᵣ) :+ (xᵢ - yᵢ)

div :: (Fractional a) => Complex a -> Complex a -> Complex a
div (xᵣ :+ xᵢ) (yᵣ :+ yᵢ) = (xᵣ * yᵣ + xᵢ * yᵢ) / d :+ (xᵢ * yᵣ - xᵣ * yᵢ) / d
  where
    d = sq yᵣ + sq yᵢ

-- helpers ---------------------------------------------------------------------

sq :: (Num a) => a -> a
sq x = x * x
